### DISCUSSION TITLE
Concatenating JPGs to PDF

### DISCUSSION THREAD

**Discussion by olin256** [ImageMagick/ImageMagick#7042] [Help]
### ImageMagick version

7.1.1-26

### Operating system, version and so on

Windows 11

### Description

I've got a bunch of JPG files of various dimensions. My goal is to create a PDF file out of all these pictures satisfying the following:

- each page is of size A4,
- each picture fits exactly onto the page and is **centered**,
- the pictures are included as they are (no resizing).

I can almost achieve this with
`convert *.jpg -page a4 output.pdf`
but the images are not centered. Adding `-gravity center` actually destroys the positioning (and it seems that the interaction of `-page` and `-gravity` is infamously obfuscated, [some say there is none](https://github.com/ImageMagick/ImageMagick/issues/2083#issuecomment-636985347), [docs say otherwise](https://imagemagick.org/script/command-line-options.php#page), the actual results are yet something different).

How can I achieve this? Note that the density might be different for each page…

**Comment by snibgo** [comment]
You say you want each page to be A4, and "each picture fits exactly onto the page". But what if the input images do not all have the same aspect ratio as A4? Then there must be blank spaces.

I suppose you want all PDF pages to have the same dimensions, width and height, in pixels. Is that correct?

Then you could `-extent` all the images to the same dimensions, eg:
```
magick *.jpg -gravity Center -extent 2100x2970 out.pdf
```
Then all the pages will be the same size, 2100x2970 pixels.

**Comment by olin256** [comment]
Sorry for not making myself clear enough. This is an example of what I want to achieve:
[w1.jpg](https://github.com/ImageMagick/ImageMagick/assets/84866288/1f91f545-52e8-451d-9271-aef7ffaefba5)
[w2.jpg](https://github.com/ImageMagick/ImageMagick/assets/84866288/1d0f5d0b-7b8e-4f35-a340-071de12cce56)
[out.pdf](https://github.com/ImageMagick/ImageMagick/files/13986462/out.pdf)

Yes, there are (necessarily) white margins. Note that

- the images are simply _included_ (no operation on them whatsoever),
- each picture is centered and spans the maximal possible area while retaining the aspect ratio (although the images have different heights),
- the PDF file correctly declares that the physical size of each page is A4.

I made this PDF file using pdfTeX, but that seems to me like a bit too heavy hammer for a simple task like this.

Specifying `-extent` doesn't stretch the images across the page (or I don't know how). Specifying `-resize` changes the images. `-page` seemed like the perfect option here, but I can't get it to center the image apart from this approach, which adds white pixels:
`magick *.jpg -gravity Center -extent 210x297# +gravity -page a4 out.pdf`
